rapids-4-spark-tools/src/main/scala/com/nvidia/spark/rapids/tool/profiling/GenerateDot.scala

```diff
@@ -0,0 +1,208 @@
+ * further.
+ *
+ * Graphviz and other tools can be used to generate images from DOT files.
+ *
+ * See https://graphviz.org/pdf/dotguide.pdf for a description of DOT files.
+ */
+object GenerateDot {
+  private val GPU_COLOR = "#76b900" // NVIDIA Green
+  private val CPU_COLOR = "#0071c5"
+  private val TRANSITION_COLOR = "red"
+
+  /**
+   * Generate a query plan visualization in dot format.
+   *
+   * @param plan First query plan and metrics
+   * @param comparisonPlan Optional second query plan and metrics
+   * @param filename Filename to write dot graph to
+   * @param includeCodegen Include WholeStageCodegen and InputAdapter nodes, if true
+   */
+  def generateDotGraph(
+      plan: QueryPlanWithMetrics,
+      comparisonPlan: Option[QueryPlanWithMetrics],
+      dir: File,
+      filename: String,
+      includeCodegen: Boolean): Unit = {
+
+    var nextId = 1
+
+    def isGpuPlan(plan: SparkPlanInfo): Boolean = {
+      plan.nodeName match {
+        case name if name contains "QueryStage" =>
+          plan.children.isEmpty || isGpuPlan(plan.children.head)
+        case name if name == "ReusedExchange" =>
+          plan.children.isEmpty || isGpuPlan(plan.children.head)
+        case name =>
+          name.startsWith("Gpu")
+      }
+    }
+
+    /**
+     * Optionally remove code-gen nodes.
+     */
+    def normalize(plan: SparkPlanInfo): SparkPlanInfo = {
+      if (!includeCodegen
+        && (plan.nodeName.startsWith("WholeStageCodegen")
+        || plan.nodeName.startsWith("InputAdapter"))) {
+        plan.children.head
+      } else {
+        plan
+      }
+    }
+
+    def formatMetric(m: SQLMetricInfo, value: Long): String = {
+      val formatter = java.text.NumberFormat.getIntegerInstance
+      m.metricType match {
+        case "timing" =>
+          val ms = value
+          s"${formatter.format(ms)} ms"
+        case "nsTiming" =>
+          val ms = TimeUnit.NANOSECONDS.toMillis(value)
+          s"${formatter.format(ms)} ms"
+        case _ =>
+          s"${formatter.format(value)}"
+      }
+    }
+
+    /** Recursively graph the operator nodes in the spark plan */
+    def writeGraph(
+        w: FileWriter,
+        node: QueryPlanWithMetrics,
+        comparisonNode: QueryPlanWithMetrics,
+        id: Int = 0): Unit = {
+
+      val nodeNormalized = normalize(node.plan)
+      val comparisonNodeNormalized = normalize(comparisonNode.plan)
+      if (nodeNormalized.nodeName == comparisonNodeNormalized.nodeName &&
+        nodeNormalized.children.length == comparisonNodeNormalized.children.length) {
+
+        val metricNames = (nodeNormalized.metrics.map(_.name) ++
+          comparisonNodeNormalized.metrics.map(_.name)).distinct.sorted
+
+        val metrics = metricNames.flatMap(name => {
+          val l = nodeNormalized.metrics.find(_.name == name)
+          val r = comparisonNodeNormalized.metrics.find(_.name == name)
+          if (l.isDefined && r.isDefined) {
```

Comment:
consider alternatives: 
- pattern match
- option.foreach or option.map
to avoid the anti-pattern if (o.defined) o.get

Reply:
Updated to use pattern matching